pkg/control/broadcast.go

```diff
@@ -0,0 +1,138 @@
+package control
+
+import (
+	context "context"
+	"crypto/tls"
+	"crypto/x509"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/horizon/pkg/grpc/lz4"
+	grpctoken "github.com/hashicorp/horizon/pkg/grpc/token"
+	"github.com/hashicorp/horizon/pkg/pb"
+	"github.com/hashicorp/horizon/pkg/utils"
+	"google.golang.org/grpc"
+	gcreds "google.golang.org/grpc/credentials"
+)
+
+type HubCatalog interface {
+	Targets() []string
+}
+
+type Broadcaster struct {
+	L       hclog.Logger
+	catalog HubCatalog
+	conn    func(addr string) (pb.HubServicesClient, error)
+}
+
+func NewBroadcaster(
+	L hclog.Logger,
+	catalog HubCatalog,
+	conn func(addr string) (pb.HubServicesClient, error),
+) (*Broadcaster, error) {
+	br := &Broadcaster{
+		L:       L,
+		catalog: catalog,
+		conn:    conn,
+	}
+
+	return br, nil
+}
+
+// AdvertiseServices gets a list of targets from the catalog and calls AddService
+// on the clients generated from the connect function (which defaults to dialing a grpc
+// connection to the target)
+func (b *Broadcaster) AdvertiseServices(ctx context.Context, as *pb.AccountServices) error {
+	var topError error
+
+	targets := b.catalog.Targets()
+
+	b.L.Info("hub broadcasting beginning", "targets", len(targets))
+
+	for _, tgt := range targets {
+		b.L.Info("broadcasting hub update", "target", tgt)
+		cli, err := b.conn(tgt)
+		if err != nil {
+			topError = multierror.Append(topError, err)
+			continue
+		}
+
+		_, err = cli.AddServices(ctx, as)
+		if err != nil {
+			topError = multierror.Append(topError, err)
+		}
+	}
+
+	return topError
+}
+
+type GRPCDial struct {
+	token string
+	cert  []byte
+
+	mu        sync.RWMutex
+	grpcConns map[string]*grpc.ClientConn
+
+	tlscfg tls.Config
+}
+
+func NewGRPCDial(token string, cert []byte) (*GRPCDial, error) {
+	g := &GRPCDial{
+		token:     token,
+		cert:      cert,
+		grpcConns: make(map[string]*grpc.ClientConn),
+	}
+
+	if g.cert != nil {
+		parsedHubCert, err := utils.ParseCertificate(cert)
+		if err != nil {
+			return nil, err
+		}
+
+		g.tlscfg.RootCAs = x509.NewCertPool()
+		g.tlscfg.RootCAs.AddCert(parsedHubCert)
+	}
+
+	return g, nil
+}
+
+func (g *GRPCDial) Dial(target string) (pb.HubServicesClient, error) {
+	g.mu.RLock()
+	cc, ok := g.grpcConns[target]
+	g.mu.RUnlock()
+
+	if ok {
+		return pb.NewHubServicesClient(cc), nil
+	}
+
+	g.mu.Lock()
```

Comment:
Why do we need to use both `RLock` and `Lock` here in sequence? I guess this relates to your comment about there being a race condition -- does the race condition require we use a `Lock` on our second check instead of an `RLock` (is there a difference in behavior that contributes to a race condition), or is the reasoning "we need to do a Write, but first we have to check that Read didn't flake out on us previously, and that check has to be within the same Lock as the eventual write-action"?

Reply:
I think of it like a table lock in SQL.  RLock locks all threads from reading, then sets the current target.  Such that everyone behind this current Dial call has to wait to read until it is done.  Only then can hypothetically the next person in line read the value, or not to see that it was changed by the person ahead of them.  But this is in the weeds for me.